ironfish/src/primitives/target.ts

```diff
@@ -140,14 +141,31 @@ export class Target {
 
     // max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
     const diffInSeconds = (time.getTime() - previousBlockTimestamp.getTime()) / 1000
-    const sign = BigInt(Math.max(1 - Math.floor(diffInSeconds / 10), -99))
-    const offset = BigInt(previousBlockDifficulty) / BigInt(DIFFICULTY_ADJUSTMENT_DENOMINATOR)
 
     // diff = parent_diff + parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
-    const difficulty = BigIntUtils.max(
-      BigInt(previousBlockDifficulty) + offset * sign,
-      Target.minDifficulty(),
-    )
+
+    const threshold = 1
+    const halfBucket = Math.floor(FREEZE_TIME_IN_SECONDS / 2)
+    const bucket =
+      Math.floor(
+        (diffInSeconds - TARGET_BLOCK_TIME_IN_SECONDS + halfBucket) / FREEZE_TIME_IN_SECONDS,
+      ) + 1
+
+    const max_change =
+      BigInt(previousBlockDifficulty) / BigInt(DIFFICULTY_ADJUSTMENT_DENOMINATOR)
+
+    let targetDifficulty = null
+    if (bucket <= threshold) {
+      // Scale up
+      const multiplier = threshold - bucket
+      targetDifficulty = previousBlockDifficulty + BigInt(max_change) * BigInt(multiplier)
+    } else {
+      // Scale down
+      const multiplier = Math.min(bucket - threshold, 99)
+      targetDifficulty = previousBlockDifficulty - BigInt(max_change) * BigInt(multiplier)
+    }
```

Comment:
what is `threshold` here? why is it 1? 

the basic idea here is if our target is 60s and our slippage is +/- 5 seconds, then first we try to figure out our bucket from target and adjust previous block's difficulty accordingly 

i think this code can be simplified quite a bit by something like: 

```
      const bucketFromTarget = findBucket(targetTime, slippage, blockTime)

      const diffInDifficulty = (parentDifficulty / BigInt(2048)) * BigInt(bucketFromTarget)

      const difficulty =
        blockTime < targetTime
          ? parentDifficulty + diffInDifficulty
          : parentDifficulty - diffInDifficulty
```